Improve typings for Angular package

waterplea opened the pull request:
There were some loose typings, as well as some misleading ones, such as config that pipe receives. Everything was updated to a stricter version, along with a small documentation change.

waterplea commented:
@lozjackson @msafi builds have been failing in this repository for nearly a year now. Please take a look at this PR since I not only improved typings but also fixed builds. It would allow other PRs to pass, if you are still interested in contributions to the project.

waterplea commented:
@themzed seeing as you became maintainer, please take a look at this PR, it fixes the build issues and has correct types which is very helpful for Angular.

themzed commented:
Hi @waterplea, thanks for PR. I can't merge to main branch right now, @msafi please take a look a this and merge if it's fine. Thanks again.

msafi commented:
Hi @themzed, I wouldn't be able to know if this is a good change or not without a closer look and manual tests. I'm afraid I don't currently have the bandwidth to do that!

waterplea commented:
@themzed @msafi How about I go over all the changes I've made? I've been using mask with Angular for over a year and built-in types are incorrect/insufficient. I've been sitting on those types all that time until recently I realized why don't I just make PR with them so everybody can use them, since that's an open source project. **Keep in mind I haven't made any changes to the logic, just the types and fixed build process**. Here are all the files I've changed:

`.travis.yml`
`angular2/karma.conf.js`
`angular2/package.json`
`angular2/tsconfig.json`
`angular2/tsconfig.test.json`
`ember/.travis.yml`

Here I've bumped node version and switched tests from ancient PhantomJS to ChromeHeadless so builds would pass. I've also changed typings file in `package.json` and `tsconfig.json` to reflect that I've moved interfaces to a separate .ts file.

`componentDocumentation.md`

Here I reflected what was missing from the docs (and also from the types) — that `config` object that pipe function receives also has `rawValue`, `previousComformedValue` and `currentCaretPosition` (when applicable).

`angular2/src/index.ts`

I've added this file since I moved all interfaces to a separate file and now we need a barrel .ts file that has everything in one place.

`angular2/src/angular2TextMask.ts`

I haven't done any changes to the code here, just moved `TextMaskConfig` interface to a separate file and expanded upon it there.

`angular2/src/interfaces.ts`

Here lies the meat of the PR. I've covered everything with jsdocs so it's easy to understand. Here's what I've changed in initial `TextMaskConfig`:

1. I've made everything `readonly` since you're not supposed to mutate configs (if you're not comfortable with that, lets remove it).
2. `TextMaskConfig` was made to interface rather than a class, there's no reason for it being a class since it doesn't have constructor or methods.
3. `mask` in function form also receives `config`, not just the `rawValue`.
4. `pipe` has corrected types — the `config` that it receives is updated with pipe specific properties and return value is not plain `object` but a proper interface with optional `indexesOfPipedChars`.
5. I've separated all complex types like `((raw: string) => Array<string | RegExp>)` into their own interfaces.

waterplea commented:
@themzed @msafi any news, guys?

waterplea commented:
@msafi looks like you guys need a new maintainer again )

msafi commented:
Although some good candidates reached out offering to maintain the project, for the security of users I can't blindly hand over maintenance tasks without my personal oversight. And I don't have interest in spending any time on this project, so I'm thinking I might announce that this project is no longer maintained and basically shut it down. The community can choose to fork and publish new packages based on it. Then users can choose to install the forked packages.

waterplea commented:
I understand how you feel. Yes, it's probably a good idea to put a disclaimer then, this seems to be THE mask package when it comes to frontend, many projects depend on it. If you could share some time to look into this PR though, you would at least leave it in a better shape for people to fork. Again, it doesn't really introduce any changes, just more precise typings and fixed builds. If not, it seems like the only way would be to go with @types package which is not recommended approach for libraries packing their own types :-(

msafi commented:
@waterplea Thank you, but merging and publishing this PR is not as straightforward for me as you think.

You've spent a lot of time on this PR, so I'm really sorry for not being able to act on it.

Please feel free to contribute the types to @types. While not the best solution, it's the quickest path at this time.
